File: problems/3Sum/threeSum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        var result = new List<IList<int>>();

        Array.Sort(nums);

        for (int pivot = 0; pivot < nums.Length; pivot++) {
            if (pivot > 0 && nums[pivot] == nums[pivot - 1]) {
                continue;
            }

            var targetSum = -(nums[pivot]);
            var startIndex = pivot + 1;
            var endIndex = nums.Length - 1;

            while (startIndex < endIndex) {
                var sumOfValuesAtStartAndEndIndex = nums[startIndex] + nums[endIndex];
                if (sumOfValuesAtStartAndEndIndex == targetSum) {
                    result.Add(new int[] { nums[pivot], nums[startIndex], nums[endIndex] });

                    while (startIndex < endIndex && nums[startIndex] == nums[startIndex + 1]) {
                        ++startIndex;
                    }

                    while (startIndex < endIndex && nums[endIndex] == nums[endIndex - 1]) {
                        --endIndex;
                    }

                    ++startIndex;
                    --endIndex;
                } else {
                    if (sumOfValuesAtStartAndEndIndex < targetSum)
                    {
                        ++startIndex;
                    } else {
                        --endIndex;
                    }
                }
            }
        }

        return result;
    }
}

File: problems/3Sum/threeSum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        HashSet<String> dedupSet = new HashSet<>();

        Arrays.sort(nums);

        for (int idx = 0; nums.length > idx; ++idx) {
            if (0 == idx || nums[idx - 1] != nums[idx]) {
                List<int[]> twoSumList = twoSum(nums, -nums[idx], 1 + idx);

                for (int[] twoSumItem : twoSumList) {
                    String key = "" + nums[idx] + nums[twoSumItem[0]] + nums[twoSumItem[1]];

                    if (!dedupSet.contains(key)) {
                        List<Integer> threeSumItem = new ArrayList<>();
                        threeSumItem.add(nums[idx]);
                        threeSumItem.add(nums[twoSumItem[0]]);
                        threeSumItem.add(nums[twoSumItem[1]]);
                        result.add(threeSumItem);
                        dedupSet.add(key);
                    }
                }
            }
        }

        return result;
    }

    public List<int[]> twoSum(int[] nums, int k, int start) {
        List<int[]> result = new ArrayList<>();
        HashMap<Integer, Integer> store = new HashMap<>();

        for (int idx = start; nums.length > idx; ++idx) {
            if (store.containsKey(k - nums[idx])) {
                result.add(new int[] {store.get(k - nums[idx]), idx});
            }

            store.put(nums[idx], idx);
        }

        return result;
    }
}
